### PR TITLE
FINOPS-119 Improve AWS Upgrade to GP3

### PR DESCRIPTION
**Pull request by douglaswth**
### Description

- Improve savings calculations based on pricing from the AWS Pricing API
- Determine Enterprise Discount Program ratio based on Optima costs or a parameter
- Support upgrading from io2 volumes as well

### Issues Resolved

FINOPS-119

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/aws/gp3_volume_upgrade/aws_upgrade_to_gp3_volume.pt [hunk 1]
```diff
@@ -133,55 +152,55 @@ datasource "ds_unattached_volumes_list" do
...
-end
-
-datasource "ds_flexera_optima" do
-  run_script $js_flexera_optima, $param_flexera_org_id_for_optima, rs_org_id
+datasource "ds_volume_types" do
+  run_script $js_volume_types, $ds_superseded_volumes
 end
 
-datasource "ds_currency_code" do
+datasource "ds_aws_products" do
```
Comment (rshade) [inline cost/aws/gp3_volume_upgrade/aws_upgrade_to_gp3_volume.pt:176]: How big is this payload?

File: cost/aws/gp3_volume_upgrade/aws_upgrade_to_gp3_volume.pt [hunk 2]
```diff
@@ -569,12 +629,24 @@ EOS
       field "size" do
         label "Size"
       end
+      field "iops" do
+        label "IOPS"
+      end
       field "tags" do
         label "Tags"
       end
       field "savings" do
```
Comment (rshade) [inline cost/aws/gp3_volume_upgrade/aws_upgrade_to_gp3_volume.pt:638]: Since we only show 6 fields by default, I think we need to move some of these up.

**Review by rshade — CHANGES_REQUESTED** [review]

File: cost/aws/gp3_volume_upgrade/aws_upgrade_to_gp3_volume.pt (hunk 1 above)
Comment (douglaswth) [reply cost/aws/gp3_volume_upgrade/aws_upgrade_to_gp3_volume.pt:176]: When it iterates over all four volume types, here's the resource usage from `fpt`:

> ### Resource Usage
>   * StartedAt: Fri, 17 Dec 2021 22:38:09 UTC
>   * FinishedAt: Fri, 17 Dec 2021 22:38:14 UTC
>   * Duration: 5.635450443s
>   * Net IO:
>     * **datasource:ds_aws_products: 0.282 mb**
>     * datasource:ds_billing_centers: 0.169 mb
>     * datasource:ds_get_caller_identity: 0.000 mb
>     * datasource:ds_regions_list: 0.004 mb
>     * datasource:ds_superseded_volumes_list: 0.007 mb
>     * datasource:ds_volume_costs: 0.001 mb
>     * ds_aws_products: 0.000 mb
>     * ds_volume_costs: 0.003 mb
>   * Mem Usage (MB):
>     * **datasource:ds_aws_products: 0.044 mb**
>     * datasource:ds_billing_centers: 0.053 mb
>     * datasource:ds_flexera_optima: 0.000 mb
>     * datasource:ds_get_caller_identity: 0.000 mb
>     * datasource:ds_regions: 0.000 mb
>     * datasource:ds_regions_list: 0.001 mb
>     * datasource:ds_superseded_volumes: 0.002 mb
>     * datasource:ds_superseded_volumes_list: 0.002 mb
>     * datasource:ds_top_level_billing_centers: 0.018 mb
>     * datasource:ds_volume_cost_mapping: 0.002 mb
>     * datasource:ds_volume_costs: 0.001 mb
>     * datasource:ds_volume_types: 0.000 mb
>   * Durations (s):
>     * **datasource:ds_aws_products: 0.246 s** (4 calls)
>     * **datasource:ds_aws_products:net_io: 0.151 s** (4 calls)
>     * **datasource:ds_aws_products:parse: 0.002 s** (4 calls)
>     * datasource:ds_billing_centers: 0.257 s (1 call)
>     * datasource:ds_billing_centers:net_io: 0.206 s (1 call)
>     * datasource:ds_billing_centers:parse: 0.002 s (1 call)
>     * datasource:ds_get_caller_identity: 0.054 s (1 call)
>     * datasource:ds_get_caller_identity:net_io: 0.035 s (1 call)
>     * datasource:ds_get_caller_identity:parse: 0.000 s (1 call)
>     * datasource:ds_regions_list: 0.100 s (1 call)
>     * datasource:ds_regions_list:net_io: 0.076 s (1 call)
>     * datasource:ds_regions_list:parse: 0.000 s (1 call)
>     * datasource:ds_superseded_volumes_list: 3.719 s (8 calls)
>     * datasource:ds_superseded_volumes_list:net_io: 3.635 s (8 calls)
>     * datasource:ds_superseded_volumes_list:parse: 0.000 s (8 calls)
>     * datasource:ds_volume_costs: 1.125 s (1 call)
>     * datasource:ds_volume_costs:net_io: 0.714 s (1 call)
>     * datasource:ds_volume_costs:parse: 0.000 s (1 call)
>     * script:js_filter_aws_superseded_volumes: 0.006 s (1 call)
>     * script:js_flexera_optima: 0.017 s (1 call)
>     * script:js_get_costs: 0.006 s (1 call)
>     * script:js_regions: 0.006 s (1 call)
>     * script:js_top_level_bc: 0.009 s (1 call)
>     * script:js_volume_cost_mapping: 0.013 s (1 call)
>     * script:js_volume_types: 0.005 s (1 call)

Things are much faster when AWS lets you filter down to almost only what you want. 😉

Comment (rshade) [reply cost/aws/gp3_volume_upgrade/aws_upgrade_to_gp3_volume.pt:176]: That is awesome, much better.

**Review by rshade — APPROVED** [review]

**Pull request merged**